<a href="https://colab.research.google.com/github/TreborRP30/An-lisis-inmobiliario-de-lima-2019/blob/main/01_Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Preparación de la mesa de trabajo**

En esta etapa se hace lo básico que es importar todas las librerías necesarias para esta limpieza de la base de datos.

In [ ]:
#IMPORTACION masiva de librerías esenciales
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

#CONFIGIRACION visual para tus gráficos (para que se vean profesionales)
sns.set_theme(style="whitegrid")
plt.rcParams['figure.figsize'] = (10, 6)

In [ ]:
#lectura
ruta = "/content/drive/MyDrive/Housing_prices_Lima_Peru.csv"
df = pd.read_csv(ruta)
print(df.shape)
df.head()

(396, 11)


,posting_id,url,precio_usd,lat,lon,distrito,area_m2,dormitorios,banos,tipo_inmueble,precio_m2
0,146884723,https://www.adondevivir.com/propiedades/clasif...,2439900.0,-12.122968,-76.970145,Santiago de Surco,2711.0,NaN,NaN,Casas,900.000000
1,149496253,https://www.adondevivir.com/propiedades/clasif...,650000.0,-12.087365,-76.948783,La Molina Vieja,620.0,3.0,2.0,Casas,1048.387097
2,149644843,https://www.adondevivir.com/propiedades/clasif...,490000.0,-12.107906,-76.981307,Chacarilla Del Estanque,312.0,5.0,4.0,Casas,1570.512821
3,146458830,https://www.adondevivir.com/propiedades/clasif...,658000.0,-12.122322,-77.037828,Miraflores,250.0,4.0,4.0,Casas,2632.000000
4,150050970,https://www.adondevivir.com/propiedades/clasif...,570000.0,-12.140899,-76.991766,Santiago de Surco,300.0,6.0,4.0,Casas,1900.000000


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 396 entries, 0 to 395
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   posting_id     396 non-null    int64  
 1   url            396 non-null    object 
 2   precio_usd     396 non-null    float64
 3   lat            396 non-null    float64
 4   lon            396 non-null    float64
 5   distrito       396 non-null    object 
 6   area_m2        395 non-null    float64
 7   dormitorios    383 non-null    float64
 8   banos          385 non-null    float64
 9   tipo_inmueble  396 non-null    object 
 10  precio_m2      395 non-null    float64
dtypes: float64(7), int64(1), object(3)
memory usage: 34.2+ KB


Realizó un análisis exploratorio de mi dataset para saber su estado, y lo que tomo en cuenta es:

•La cantidad de columnas y filas: 396 columnas y 11 filas.

•Los tipos de datos de las columnas: 7 decimales, 1 enteros y 3 categoricas.

•Si hay celdas vacias: hay algunas filas que si tienen celdas vacias y voy a eliminarla para que no allá alteración de los datos. En mi caso es permisible eliminar filas con celda vacías por el tamaño de el data set

In [ ]:
df.dropna(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 378 entries, 1 to 395
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   posting_id     378 non-null    int64  
 1   url            378 non-null    object 
 2   precio_usd     378 non-null    float64
 3   lat            378 non-null    float64
 4   lon            378 non-null    float64
 5   distrito       378 non-null    object 
 6   area_m2        378 non-null    float64
 7   dormitorios    378 non-null    float64
 8   banos          378 non-null    float64
 9   tipo_inmueble  378 non-null    object 
 10  precio_m2      378 non-null    float64
dtypes: float64(7), int64(1), object(3)
memory usage: 35.4+ KB


🛠️ **Fase de Limpieza Estructural:**

Optimizando el Dataset Inmobiliario
En la etapa actual de extracción y limpieza (ETL) con Python, el objetivo principal es garantizar que nuestro dataset sea eficiente, escalable y esté listo para el modelado de datos. Tras una evaluación analítica de las variables disponibles, he aplicado decisiones estratégicas sobre la estructura de la información:


🗑️ Variables Eliminadas
url (Enlace de la publicación): He decidido prescindir de esta columna para optimizar el rendimiento y el peso del dataset. Al ser cadenas de texto extensas, consumen memoria innecesaria y no aportan valor estadístico ni descriptivo para responder a la pregunta central de nuestro negocio (Rentabilidad y ROI).


🛡️ Variables Conservadas y su Propósito Estratégico
He mantenido las siguientes columnas clave pensando en la arquitectura técnica y los objetivos de visualización del proyecto:


🔑 posting_id (Identificador Único): Es el pilar de nuestra arquitectura de datos. Esta columna será nuestra Llave Primaria (Primary Key) cuando migremos la información a SQL, permitiéndonos construir un modelo relacional robusto (Esquema Estrella) para cruzar estas propiedades con futuras tablas de dimensiones.


🗺️ lat y lon (Coordenadas Espaciales): Son fundamentales para el análisis geoespacial. Las conservo porque el precio en un distrito puede variar drásticamente según la subzona. Estas variables nos permitirán construir mapas de calor e interactivos en Power BI, y analizar la cercanía a factores externos (vías principales, transporte).


💰 precio_m2 (Valor por Metro Cuadrado): Es la métrica financiera core para nuestro análisis exploratorio. Aunque podría calcularse posteriormente, traerla estandarizada desde Python agiliza la comparación del valor real del suelo entre distintos distritos y es vital para el cálculo de la rentabilidad.

In [ ]:
#ELIMINAR columnas basura
columnas_basura = ['url']
df.drop(columns=columnas_basura, inplace=True, errors='ignore')

#VERIFICAMOS la información
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 378 entries, 1 to 395
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   posting_id     378 non-null    int64  
 1   precio_usd     378 non-null    float64
 2   lat            378 non-null    float64
 3   lon            378 non-null    float64
 4   distrito       378 non-null    object 
 5   area_m2        378 non-null    float64
 6   dormitorios    378 non-null    float64
 7   banos          378 non-null    float64
 8   tipo_inmueble  378 non-null    object 
 9   precio_m2      378 non-null    float64
dtypes: float64(7), int64(1), object(2)
memory usage: 32.5+ KB
None


🧹 **Tratamiento Híbrido de Outliers: Combinando Estadística y Reglas de Negocio**


Durante la fase de limpieza (ETL) en Python de mi proyecto de Análisis Inmobiliario en Lima, me topé con un desafío clásico: la presencia de valores atípicos (outliers). Al explorar los datos, identifiqué registros con precios absurdos de $1 o propiedades listadas con hasta 19 habitaciones, errores que sesgarían por completo cualquier cálculo de rentabilidad y ROI.


Para limpiar el dataset sin correr el riesgo de eliminar propiedades legítimas, decidí diseñar e implementar un enfoque híbrido utilizando máscaras booleanas en Pandas:


📈 1. Filtro Estadístico (Rango Intercuartílico - IQR)
Para las variables financieras y de escala continua (precio_usd, area_m2 y precio_m2), implementé el método matemático IQR. Mediante un bucle for, el algoritmo evalúa estas métricas y recorta los extremos de la distribución. Esto limpia automáticamente errores de digitación en precios y descarta terrenos de escala industrial que no aportan a mi análisis residencial.


🧠 2. Filtro por Reglas de Negocio
Aplicar estadística pura a variables discretas (como el número de dormitorios o banos) habría sido un error, ya que el IQR podría considerar una casa legítima de 5 habitaciones como una "anomalía". Para evitar esta pérdida de datos valiosos, apliqué reglas de negocio basadas en la lógica del mercado, limitando estos espacios a un rango residencial realista (de 1 a 6).


🛠️ El Resultado
Al interceptar ambas máscaras lógicas, logré depurar el dataset con precisión. El resultado es una base de datos robusta, íntegra y libre de ruido. Ahora los datos están listos para ser modelados en SQL y, finalmente, dar vida a mi dashboard interactivo desarrollado netamente con librerías de Python.

In [ ]:
#Definimos las variables numéricas más críticas de tu dataset
columnas_clave = ['precio_usd', 'area_m2', 'dormitorios', 'banos','precio_m2']

#Generamos un resumen estadístico, mostrando los valores mínimos y máximos
df[columnas_clave].describe().round(2)

,precio_usd,area_m2,dormitorios,banos,precio_m2
count,378.00,378.00,378.00,378.00,378.00
mean,759121.43,585.87,4.94,4.09,1649.70
std,831526.97,867.69,2.24,1.94,946.32
min,1.00,26.00,1.00,1.00,0.00
25%,320000.00,221.00,4.00,3.00,1017.18
50%,550000.00,327.50,4.00,4.00,1529.51
75%,876750.00,643.25,5.00,5.00,1994.80
max,9500000.00,10656.00,19.00,20.00,7371.79


In [ ]:
#Filtro Estadístico (IQR) para variables continuas
columnas_iqr = ['precio_usd', 'area_m2', 'precio_m2']
mascara_iqr = pd.Series(True, index=df.index)

for col in columnas_iqr:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1

    limite_inferior = Q1 - 1.5 * IQR
    limite_superior = Q3 + 1.5 * IQR

    mascara_iqr = mascara_iqr & (df[col] >= limite_inferior) & (df[col] <= limite_superior)

#Filtro por Reglas de Negocio para dormitorios y baños
#Así eliminamos ese outlier de 19 cuartos y 20 baños, pero conservamos propiedades grandes reales
mascara_logica = (
    (df['dormitorios'] >= 1) & (df['dormitorios'] <= 6) &
    (df['banos'] >= 1) & (df['banos'] <= 6)
)

#Combinamos ambas condiciones en un nuevo DataFrame limpio
df_limpio = df[mascara_iqr & mascara_logica]

#Evaluamos el impacto del proceso de limpieza
filas_eliminadas = df.shape[0] - df_limpio.shape[0]
print(f"📋 Filas iniciales: {df.shape[0]}")
print(f"✨ Filas finales tras limpieza total: {df_limpio.shape[0]}")
print(f"🚫 Total de outliers detectados y eliminados: {filas_eliminadas}")

🏗️ **Estandarización de Variables Categóricas y Reducción de Dimensionalidad**

1. Normalización Geoespacial (distrito):
Al auditar la variable distrito, se detectó una mezcla de jerarquías: el scraper extrajo nombres de urbanizaciones y subzonas (ej. Camacho, Las Casuarinas) tratándolas como distritos independientes. Se implementó un diccionario de mapeo para reasignar más de 50 subzonas a sus distritos oficiales correspondientes, garantizando agrupaciones limpias y precisas para la fase de visualización.

2. Depuración por Varianza Nula (tipo_inmueble):
Dado que el alcance de este proyecto está estratégicamente acotado al segmento de Casas Independientes para evitar sesgos en el cálculo del ROI frente a departamentos, la variable tipo_inmueble presenta una varianza de cero (100% de los registros son 'Casas'). Al carecer de poder predictivo o descriptivo adicional, el algoritmo automatizado de calidad de datos procede a eliminar esta columna para optimizar la estructura del dataset.

In [ ]:
#Inspeccionando las variables categóricas en busca de errores
print("--- Distritos Únicos ---")
print(df_limpio['distrito'].unique())

print("\n--- Tipos de Inmueble Únicos ---")
print(df_limpio['tipo_inmueble'].unique())

--- Distritos Únicos ---
['La Molina Vieja' 'Chacarilla Del Estanque' 'Miraflores'
 'Santiago de Surco' 'El Remanso' 'Los Precursores' 'San Isidro'
 'La Molina' 'El Sol De La Molina Et.1' 'Rinconada Del Lago'
 'Los Granados' 'Campo Verde' 'La Planicie' 'Aurora' 'Chorrillos'
 'San Borja' 'La Encantada De Villa' 'La Fontana' 'Capebco' 'San Miguel'
 'Villa Club' 'El Sol De La Molina Et.2' 'Rinconada Baja' 'Corpac'
 'Centro Comercial Monterrico' 'San Borja Norte' 'San Martín de Porres'
 'Santa Cruz' 'Orrantia Del Mar' 'Puente Piedra' 'Ate Vitarte'
 'Villa María del Triunfo' 'Comas' 'Limatambo' 'Los Rosales' 'Orrantia'
 'Lima' 'Neptuno' 'Monterrico Chico' 'San Juan de Miraflores'
 'Los Cedros De Villa' 'Camacho' 'Las Lomas De La Molina Vieja'
 'Villa Sol' 'San Juan de Lurigancho' 'Carabayllo' 'Lurín' 'Magdalena'
 'Pachacamac' 'Lima Cercado' 'San Borja Sur' 'Valle Hermoso'
 'Alameda Del Pinar' 'Santa Teresa' 'Las Magnolias' 'Santa Monica'
 'Pueblo Libre' 'Villa el Salvador' 'Las Casuarinas' 

In [ ]:
#Creamos el diccionario de mapeo de subzonas a distritos oficiales de Lima
correccion_distritos = {
    #La Molina y sus urbanizaciones
    'La Molina Vieja': 'La Molina',
    'El Remanso': 'La Molina',
    'El Sol De La Molina Et.1': 'La Molina',
    'El Sol De La Molina Et.2': 'La Molina',
    'El Sol De La Molina Et.3': 'La Molina',
    'Rinconada Del Lago': 'La Molina',
    'Rinconada Baja': 'La Molina',
    'Rinconada Del Lago Etapa II': 'La Molina',
    'La Planicie': 'La Molina',
    'Portada De La Planicie': 'La Molina',
    'La Planicie Zona Este': 'La Molina',
    'La Fontana': 'La Molina',
    'Camacho': 'La Molina',
    'Las Lomas De La Molina Vieja': 'La Molina',
    'Las Lomas De La Molina Vieja Y II': 'La Molina',
    'Las Lomas De La Molina Vieja E I': 'La Molina',
    'Santa Patricia I-etapa': 'La Molina',
    'Sirius': 'La Molina',
    'El Mastil De La Laguna': 'La Molina',
    'El Refugio': 'La Molina',

    #Santiago de Surco y sus urbanizaciones
    'Chacarilla Del Estanque': 'Santiago de Surco',
    'Centro Comercial Monterrico': 'Santiago de Surco',
    'Monterrico Chico': 'Santiago de Surco',
    'Los Alamos De Monterrico': 'Santiago de Surco',
    'Los Precursores': 'Santiago de Surco',
    'Limatambo': 'Santiago de Surco',
    'Las Torres de Limatambo': 'Santiago de Surco',
    'Los Rosales': 'Santiago de Surco',
    'Neptuno': 'Santiago de Surco',
    'Valle Hermoso': 'Santiago de Surco',
    'Las Casuarinas': 'Santiago de Surco',
    'Casuarinas Sur': 'Santiago de Surco',
    'Las Gardenias': 'Santiago de Surco',
    'Santa Teresa': 'Santiago de Surco',
    'Prolongacion Benavides': 'Santiago de Surco',

    #San Borja y San Isidro
    'San Borja Norte': 'San Borja',
    'San Borja Sur': 'San Borja',
    'Corpac': 'San Isidro',
    'Orrantia Del Mar': 'San Isidro',
    'Orrantia': 'San Isidro',
    'Golf': 'San Isidro',
    'Country Club': 'San Isidro',
    'El Olivar': 'San Isidro',

    #Miraflores
    'Aurora': 'Miraflores',
    'Santa Cruz': 'Miraflores',
    'Armendariz': 'Miraflores',

    #Chorrillos
    'La Encantada De Villa': 'Chorrillos',
    'Los Cedros de Villa': 'Chorrillos',
    'Ariosto Matellini': 'Chorrillos',
    'La Campiña': 'Chorrillos',

    #Otros distritos (Norte, Sur, Centro)
    'Maranga': 'San Miguel',
    'Las Leyendas': 'San Miguel',
    'Ate Vitarte': 'Ate',
    'Zarate': 'San Juan de Lurigancho',
    'Villa Sol': 'Los Olivos',
    'Las Palmeras': 'Los Olivos',
    'Alameda Del Pinar': 'Comas',
    'Lima Cercado': 'Lima',
    'Magdalena': 'Magdalena del Mar',

    #Balnearios y Sur Chico
    'Paso Chico': 'Lurin',
    'Islas Del Sol': 'Lurin',
    'Arica': 'Lurin',
    'Pulpos': 'Lurin',
    'Caballeros': 'Punta Hermosa',
}

#Aplicamos el reemplazo.
# Los distritos que ya estaban bien (ej. 'San Miguel') no se verán afectados.
df_limpio['distrito'] = df_limpio['distrito'].replace(correccion_distritos)

#Verificamos el resultado final de nuestras agrupaciones
print(f"Total de distritos únicos tras limpieza: {df_limpio['distrito'].nunique()}")
print(sorted(df_limpio['distrito'].unique()))

Total de distritos únicos tras limpieza: 50
['Ate', 'Barranco', 'Campo Verde', 'Capebco', 'Carabayllo', 'Chorrillos', 'Cieneguilla', 'Comas', 'El Agustino', 'Jesús María', 'La Alborada', 'La Almudena', 'La Molina', 'Las Lomas De La Molina Vieja E Ii', 'Las Magnolias', 'Las Torres De Limatambo', 'Lima', 'Los Cedros De Villa', 'Los Cipreses', 'Los Gladiolos', 'Los Granados', 'Los Olivos', 'Los Sauces', 'Los Viñedos', 'Lurin', 'Lurín', 'Magdalena del Mar', 'Miraflores', 'Pachacamac', 'Pueblo Libre', 'Puente Piedra', 'Punta Hermosa', 'Punta Negra', 'Rinconada Del Lago Etapa Ii', 'San Borja', 'San Eugenio', 'San Isidro', 'San Juan de Lurigancho', 'San Juan de Miraflores', 'San Martín de Porres', 'San Miguel', 'Santa Anita', 'Santa Isabel', 'Santa Monica', 'Santa Rosa Alta', 'Santa Rosita', 'Santiago de Surco', 'Villa Club', 'Villa María del Triunfo', 'Villa el Salvador']


/tmp/ipykernel_688/1298956903.py:84: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_limpio['distrito'] = df_limpio['distrito'].replace(correccion_distritos)


🚀 **El Resultado Final**

Con este paso, declaro oficialmente terminada la fase de limpieza estructural (ETL). He transformado datos crudos y ruidosos en un dataset robusto, íntegro y estandarizado. El terreno está completamente preparado para la siguiente fase: dar vida a los datos mediante un análisis exploratorio y empezará contruir en dashboard en python. 🐍📊

In [ ]:
#DEFINIMOS la ruta exacta dentro de tu Drive.
ruta_drive = '/content/drive/MyDrive/casas_lima_limpio.csv'

#EXPORTAMOS el DataFrame a CSV directamente a tu nube
df_limpio.to_csv(ruta_drive,
                 index=False,       #EVITAMOS la columna extra de numeración
                 encoding='utf-8')  #PROTEGEMOS las tildes y caracteres especiales

print("✅ ¡Exportación a Google Drive exitosa!")
print(f"📁 Tu archivo está a salvo en: {ruta_drive}")